druid-shell/src/platform/windows/application.rs

```diff
@@ -17,52 +17,86 @@
 use std::mem;
 use std::ptr;
 
+use winapi::shared::minwindef::FALSE;
 use winapi::shared::minwindef::HINSTANCE;
 use winapi::shared::ntdef::LPCWSTR;
 use winapi::shared::windef::HCURSOR;
+use winapi::shared::winerror::HRESULT_FROM_WIN32;
+use winapi::um::errhandlingapi::GetLastError;
 use winapi::um::shellscalingapi::PROCESS_SYSTEM_DPI_AWARE;
 use winapi::um::wingdi::CreateSolidBrush;
 use winapi::um::winuser::{
-    DispatchMessageW, GetAncestor, GetMessageW, LoadIconW, PostQuitMessage, RegisterClassW,
-    TranslateAcceleratorW, TranslateMessage, GA_ROOT, IDI_APPLICATION, MSG, WNDCLASSW,
+    DispatchMessageW, GetAncestor, GetMessageW, LoadIconW, PostQuitMessage, PostThreadMessageW,
+    RegisterClassW, TranslateAcceleratorW, TranslateMessage, GA_ROOT, IDI_APPLICATION, MSG,
+    WNDCLASSW,
 };
 
+use log::{error, warn};
+
 use crate::application::AppHandler;
 
 use super::accels;
 use super::clipboard::Clipboard;
-use super::util::{self, ToWide, CLASS_NAME, OPTIONAL_FUNCTIONS};
-use super::window::win_proc_dispatch;
+use super::error::Error;
+use super::util::{
+    self, claim_main_thread, main_thread_id, release_main_thread, ToWide, CLASS_NAME,
+    OPTIONAL_FUNCTIONS,
+};
+use super::window::{win_proc_dispatch, DS_REQUEST_QUIT};
 
-pub struct Application;
+pub struct Application {
+    handler: Option<Box<dyn AppHandler>>,
+}
 
 impl Application {
-    pub fn new(_handler: Option<Box<dyn AppHandler>>) -> Application {
+    pub fn new(handler: Option<Box<dyn AppHandler>>) -> Application {
         Application::init();
-        Application
+        Application { handler }
     }
 
     pub fn run(&mut self) {
+        claim_main_thread();
```

Comment:
I'm a little unclear what the motivation for this is; it seems orthogonal to quit logic. I'm also concerned about safety implications if the main thread is changed; right now we don't require `Sync` or `Send` on the wrapped wndproc, but if the wndproc can be called from a different thread than the one it was created on, it could violate soundness.

Reply:
The motivation is to know the thread id where to send `DS_REQUEST_QUIT` (or even `PostQuitMessage` for completeness). The `Application::quit()` function is just a static associated function and operates on global state.

Reply:
Ah, I see. I was confused by the naming; I thought it was changing what Windows considered the main thread, but on re-read I see that isn't the case.

Reply:
Yeah I just mimicked what the other platforms do with `util::assert_main_thread()`.